File: structures/stack/implementation/stack.java
import java.util.Stack;

public class Main {
    
    // Function to push an element onto the stack (avec le message d'affichage du C)
    public static void push(Stack<Integer> stack, int value) {
        stack.push(value);
        System.out.println("Pushed " + value + " onto the stack");
    }

    // Function to pop an element from the stack
    public static int pop(Stack<Integer> stack) {
        if (stack.isEmpty()) {
            System.out.println("Stack Underflow");
            return -1;
        }
        int popped = stack.pop();
        System.out.println("Popped " + popped + " from the stack");
        return popped;
    }

    // Function to peek the top element of the stack
    public static int peek(Stack<Integer> stack) {
        if (stack.isEmpty()) {
            System.out.println("Stack is empty");
            return -1;
        }
        return stack.peek();
    }

    // Driver Code
    public static void main(String[] args) {
        // Utilisation directe de la pile native de Java
        Stack<Integer> stack = new Stack<>();

        push(stack, 3);
        System.out.println("Top element: " + peek(stack));

        push(stack, 5);
        System.out.println("Top element: " + peek(stack));

        push(stack, 2);
        System.out.println("Top element: " + peek(stack));

        push(stack, 8);
        System.out.println("Top element: " + peek(stack));

        while (!stack.isEmpty()) {
            System.out.println("Top element: " + peek(stack));
            pop(stack);
        }
    }
}

File: structures/stack/implementation/stack.py
# Python Implementation of a Stack using an Array / Fixed Capacity

MAX_SIZE = 100

class Stack:
    def __init__(self):
        # Initialisation du tableau avec une taille fixe et de top à -1
        self.arr = [0] * MAX_SIZE
        self.top = -1

# Function to initialize the stack (géré par le constructeur en Python, mais recréé pour correspondre au C)
def initialize(stack):
    stack.top = -1

# Function to check if the stack is empty
def isEmpty(stack):
    return stack.top == -1

# Function to check if the stack is full
def isFull(stack):
    return stack.top >= MAX_SIZE - 1

# Function to push an element onto the stack
def push(stack, value):
    if isFull(stack):
        print("Stack Overflow")
        return
    stack.top += 1
    stack.arr[stack.top] = value
    print(f"Pushed {value} onto the stack")

# Function to pop an element from the stack
def pop(stack):
    if isEmpty(stack):
        print("Stack Underflow")
        return -1

    popped = stack.arr[stack.top]
    stack.top -= 1
    print(f"Popped {popped} from the stack")
    return popped

# Function to peek the top element of the stack
def peek(stack):
    if isEmpty(stack):
        print("Stack is empty")
        return -1
    return stack.arr[stack.top]

# Main
if __name__ == "__main__":
    stack = Stack()
    initialize(stack)

    push(stack, 3)
    print(f"Top element: {peek(stack)}")

    push(stack, 5)
    print(f"Top element: {peek(stack)}")

    push(stack, 2)
    print(f"Top element: {peek(stack)}")

    push(stack, 8)
    print(f"Top element: {peek(stack)}")

    while not isEmpty(stack):
        print(f"Top element: {peek(stack)}")
        print(f"Popped element: {pop(stack)}")
